File: ansible_appview/phoenix/lib/ansible_appview/protocol.ex
defmodule AnsibleAppview.Protocol do
  @moduledoc """
  App↔AppView protocol version constants (service architecture plan, Phase 0).

  Clients advertise the protocol version they speak via the
  `x-ansible-protocol` request header; the AppView advertises its current and
  minimum supported versions in `GET /api/v1/meta`. Mirrors
  `AnsibleRelay.Protocol`.
  """

  @current_version 1
  @min_supported_version 1
  @header "x-ansible-protocol"

  @doc "Highest protocol version this AppView speaks."
  def current_version, do: @current_version

  @doc "Oldest client protocol version still accepted."
  def min_supported_version, do: @min_supported_version

  @doc "Request header carrying the client's protocol version."
  def header, do: @header

  @doc "Version block advertised in meta/discovery responses."
  def advertisement do
    %{current: @current_version, min_supported: @min_supported_version}
  end
end

File: ansible_node/app/lib/config/protocol.dart
/// App↔relay/appview protocol version constants (service architecture plan,
/// Phase 0 — API versioning). Mirrors `AnsibleRelay.Protocol` and
/// `AnsibleAppview.Protocol`.
///
/// Every relay/appview HTTP client sends [headers] so servers can enforce a
/// minimum supported version. A `426 {"error": "upgrade_required"}` response
/// means this build is older than the server's minimum and the user must
/// update the app (mapped to friendly copy in lib/l10n/user_facing_error.dart).
class AnsibleProtocol {
  const AnsibleProtocol._();

  /// Highest protocol version this app speaks. Ops with a `schema_version`
  /// above this are unknown future formats and are skipped during sync
  /// instead of being misparsed.
  static const int currentVersion = 1;

  /// Request header carrying the client's protocol version.
  static const String headerName = 'x-ansible-protocol';

  /// Spread into the request headers of every relay/appview HTTP client.
  static const Map<String, String> headers = {headerName: '$currentVersion'};

  /// Server error code accompanying an HTTP 426 when this client is older
  /// than the server's minimum supported protocol version.
  static const String upgradeRequiredCode = 'upgrade_required';
}
